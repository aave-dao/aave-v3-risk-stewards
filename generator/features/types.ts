import {NumberInputValues, PercentInputValues} from '../prompts';

export interface AssetSelector {
  asset: string;
}

export interface CapsUpdatePartial {
  supplyCap: NumberInputValues;
  borrowCap: NumberInputValues;
}

export interface CapsUpdate extends CapsUpdatePartial, AssetSelector {}

export interface CollateralUpdatePartial {
  ltv: PercentInputValues;
  liqThreshold: PercentInputValues;
  liqBonus: PercentInputValues;
  debtCeiling: NumberInputValues;
  liqProtocolFee: PercentInputValues;
}

export interface CollateralUpdate extends CollateralUpdatePartial, AssetSelector {}

export interface RateStrategyParams {
  optimalUtilizationRate: string;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;
  optimalStableToTotalDebtRatio?: string;
}

export interface RateStrategyUpdate extends AssetSelector {
  params: RateStrategyParams;
}

export interface OracleSelector {
  oracle: string;
}

export interface LstPriceCapUpdatePartial {
  snapshotTimestamp: NumberInputValues;
  snapshotRatio: NumberInputValues;
  maxYearlyRatioGrowthPercent: PercentInputValues;
}

export interface StablePriceCapUpdatePartial {
  priceCap: NumberInputValues;
}

export interface LstPriceCapUpdate extends LstPriceCapUpdatePartial, AssetSelector {}

export interface StablePriceCapUpdate extends StablePriceCapUpdatePartial, AssetSelector {}
