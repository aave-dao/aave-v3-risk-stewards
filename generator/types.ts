import * as addressBook from '@bgd-labs/aave-address-book';
import {
  CapsUpdate,
  CollateralUpdate,
  LstPriceCapUpdate,
  RateStrategyUpdate,
  StablePriceCapUpdate,
} from './features/types';

export const V3_POOLS = [
  'AaveV3Ethereum',
  'AaveV3EthereumLido',
  'AaveV3Polygon',
  'AaveV3Avalanche',
  'AaveV3Optimism',
  'AaveV3Arbitrum',
  'AaveV3Metis',
  'AaveV3Base',
  'AaveV3Gnosis',
  'AaveV3Scroll',
  'AaveV3BNB'
] as const satisfies readonly (keyof typeof addressBook)[];

export const POOLS = [
  ...V3_POOLS,
] as const satisfies readonly (keyof typeof addressBook)[];

export type PoolIdentifier = (typeof POOLS)[number];
export type PoolIdentifierV3 = (typeof V3_POOLS)[number];

export interface Options {
  force?: boolean;
  pools: PoolIdentifier[];
  title: string;
  // automatically generated shortName from title
  shortName: string;
  author: string;
  discussion: string;
  configFile?: string;
  date: string;
}

export type PoolConfigs = Partial<Record<PoolIdentifier, PoolConfig>>;

export type CodeArtifact = {
  code?: {
    constants?: string[];
    fn?: string[];
    execute?: string[];
  };
};

export enum FEATURE {
  CAPS_UPDATE = 'CAPS_UPDATE',
  COLLATERALS_UPDATE = 'COLLATERALS_UPDATE',
  RATE_UPDATE_V3 = 'RATE_UPDATE_V3',
  LST_PRICE_CAP_UPDATE = 'LST_PRICE_CAP_UPDATE',
  STABLE_PRICE_CAP_UPDATE = 'STABLE_PRICE_CAP_UPDATE',
}

export interface FeatureModule<T extends {} = {}> {
  description: string;
  value: FEATURE;
  cli: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache}) => Promise<T>;
  build: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache; cfg: T}) => CodeArtifact;
}

export const ENGINE_FLAGS = {
  KEEP_CURRENT: 'KEEP_CURRENT',
  KEEP_CURRENT_STRING: 'KEEP_CURRENT_STRING',
  KEEP_CURRENT_ADDRESS: 'KEEP_CURRENT_ADDRESS',
  ENABLED: 'ENABLED',
  DISABLED: 'DISABLED',
} as const;

export const AVAILABLE_VERSIONS = {V3: 'V3'} as const;

export type ConfigFile = {
  rootOptions: Options;
  poolOptions: Partial<Record<PoolIdentifier, Omit<PoolConfig, 'artifacts'>>>;
};

export type PoolCache = {blockNumber: number};

export interface PoolConfig {
  artifacts: CodeArtifact[];
  configs: {
    [FEATURE.CAPS_UPDATE]?: CapsUpdate[];
    [FEATURE.COLLATERALS_UPDATE]?: CollateralUpdate[];
    [FEATURE.RATE_UPDATE_V3]?: RateStrategyUpdate[];
    [FEATURE.LST_PRICE_CAP_UPDATE]?: LstPriceCapUpdate[];
    [FEATURE.STABLE_PRICE_CAP_UPDATE]?: StablePriceCapUpdate[];
  };
  cache: PoolCache;
}

export type Files = {
  jsonConfig: string;
  payloads: {pool: PoolIdentifier; payload: string; contractName: string}[];
};
