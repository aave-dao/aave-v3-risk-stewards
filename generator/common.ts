import * as addressBook from '@bgd-labs/aave-address-book';
import {Options, PoolIdentifier, PoolIdentifierV3} from './types';
import {
  arbitrum,
  avalanche,
  mainnet,
  metis,
  optimism,
  polygon,
  base,
  bsc,
  gnosis,
  scroll,
  zkSync,
  linea,
  sonic,
  celo,
  soneium,
  plasma,
} from 'viem/chains';

export const AVAILABLE_CHAINS = [
  'Ethereum',
  'Optimism',
  'Arbitrum',
  'Polygon',
  'Avalanche',
  'Fantom',
  'Harmony',
  'Metis',
  'Base',
  'BNB',
  'Gnosis',
  'Scroll',
  'ZkSync',
  'Linea',
  'Sonic',
  'Celo',
  'Soneium',
  'Plasma',
] as const;

export function getAssets(pool: PoolIdentifier): string[] {
  const assets = addressBook[pool].ASSETS;
  return Object.keys(assets);
}

export function getEModes(pool: PoolIdentifierV3): {value: string; id: number}[] {
  return Object.keys(addressBook[pool].E_MODES).map((key) => ({
    // map the complex type to a string as used in the sol libs
    value: addressBook[pool].E_MODES[key].label.toUpperCase().replace(/[^A-Z0-9]+/gi, '_'),
    id: key as unknown as number,
  }));
}

export function getVersion(pool: PoolIdentifier) {
  return 'V3';
}

export function getPoolChain(pool: PoolIdentifier) {
  const chain = AVAILABLE_CHAINS.find((chain) => pool.indexOf(chain) !== -1);
  if (!chain) throw new Error('cannot find chain for pool');
  return chain;
}

export function getPoolSuffix(pool: PoolIdentifier) {
  return pool.replace('AaveV3', '');
}

export function getDate() {
  const date = new Date();
  const years = date.getFullYear();
  const months = date.getMonth() + 1; // it's js so months are 0 indexed
  const day = date.getDate();
  return `${years}${months <= 9 ? '0' : ''}${months}${day <= 9 ? '0' : ''}${day}`;
}

/**
 * Prefix with the date for proper sorting
 * @param {*} options
 * @returns
 */
export function generateFolderName(options: Options) {
  return `${options.date}_${options.pools.length === 1 ? options.pools[0] : 'Multi'}_${
    options.shortName
  }`;
}

/**
 * Suffix with the date as prefixing would generate invalid contract names
 * @param {*} options
 * @param {*} chain
 * @returns
 */
export function generateContractName(options: Options, pool?: PoolIdentifier) {
  let name = pool ? `${pool}_` : '';
  name += `${options.shortName}`;
  name += `_${options.date}`;
  return name;
}

export function getChainAlias(chain) {
  return chain === 'Ethereum' ? 'mainnet' : chain.toLowerCase();
}

export function pascalCase(str: string) {
  return str
    .replace(/[\W]/g, ' ') // remove special chars as this is used for solc contract name
    .replace(/(\w)(\w*)/g, function (g0, g1, g2) {
      return g1.toUpperCase() + g2;
    })
    .replace(/ /g, '');
}

export const CHAIN_TO_CHAIN_ID = {
  Ethereum: mainnet.id,
  Polygon: polygon.id,
  Optimism: optimism.id,
  Arbitrum: arbitrum.id,
  Avalanche: avalanche.id,
  Metis: metis.id,
  Base: base.id,
  BNB: bsc.id,
  Gnosis: gnosis.id,
  Scroll: scroll.id,
  ZkSync: zkSync.id,
  Linea: linea.id,
  Sonic: sonic.id,
  Celo: celo.id,
  Soneium: soneium.id,
  Plasma: plasma.id,
};

export function flagAsRequired(message: string, required?: boolean) {
  return required ? `${message}*` : message;
}
