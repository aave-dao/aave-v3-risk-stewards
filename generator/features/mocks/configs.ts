import {Options} from '../../types';
import {CapsUpdate, CollateralUpdate, RateStrategyUpdate, LstPriceCapUpdate, StablePriceCapUpdate, EModeCategoryUpdate} from '../types';

export const MOCK_OPTIONS: Options = {
  pools: ['AaveV3Ethereum'],
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  author: 'test',
  discussion: 'test',
};

export const capUpdate: CapsUpdate[] = [
  {
    asset: 'WETH',
    supplyCap: '10000',
    borrowCap: '5000'
  },
];

export const collateralUpdate: CollateralUpdate[] = [
  {
    asset: 'DAI',
    ltv: '8500',
    liqThreshold: '8800',
    liqBonus: '600',
    debtCeiling: '',
    liqProtocolFee: '1200'
  },
];

export const eModeUpdate: EModeCategoryUpdate[] = [
  {
    eModeCategory: 'AaveV3EthereumEModes.ETH_CORRELATED',
    ltv: '',
    liqThreshold: '50',
    liqBonus: '',
    label: '',
  },
];

export const rateUpdateV3: RateStrategyUpdate[] = [
  {
    asset: 'WETH',
    params: {
      optimalUtilizationRate: '',
      baseVariableBorrowRate: '6',
      variableRateSlope1: '',
      variableRateSlope2: '',
    },
  },
  {
    asset: 'DAI',
    params: {
      optimalUtilizationRate: '',
      baseVariableBorrowRate: '4',
      variableRateSlope1: '10',
      variableRateSlope2: '',
    },
  },
  {
    asset: 'USDC',
    params: {
      optimalUtilizationRate: '',
      baseVariableBorrowRate: '4',
      variableRateSlope1: '10',
      variableRateSlope2: '',
    },
  },
  {
    asset: 'USDT',
    params: {
      optimalUtilizationRate: '',
      baseVariableBorrowRate: '6',
      variableRateSlope1: '10',
      variableRateSlope2: '',
    },
  },
  {
    asset: 'WBTC',
    params: {
      optimalUtilizationRate: '',
      baseVariableBorrowRate: '5',
      variableRateSlope1: '',
      variableRateSlope2: '',
    },
  },
];

export const lstPriceCapUpdate: LstPriceCapUpdate[] = [
  {
    asset: 'wstETH',
    snapshotTimestamp: '1723621200',
    snapshotRatio: '1177101458282319168',
    maxYearlyRatioGrowthPercent: '10.64',
  },
];

export const stablePriceCapUpdate: StablePriceCapUpdate[] = [
  {
    asset: 'USDT',
    priceCap: '108000000'
  }
];
