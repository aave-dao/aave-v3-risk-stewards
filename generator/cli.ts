import path from 'path';
import {Command, Option} from 'commander';
import {CHAIN_TO_CHAIN_ID, getDate, getPoolChain, pascalCase} from './common';
import {input, checkbox} from '@inquirer/prompts';
import {
  ConfigFile,
  Options,
  POOLS,
  PoolCache,
  PoolConfigs,
  PoolIdentifier,
} from './types';
import {capsUpdates} from './features/capsUpdates';
import {rateUpdatesV3} from './features/rateUpdates';
import {collateralsUpdates} from './features/collateralsUpdates';
import {eModeUpdates} from './features/eModeUpdates';
import {lstPriceCapsUpdates} from './features/lstPriceCapsUpdates';
import {stablePriceCapsUpdates} from './features/stablePriceCapsUpdates';
import {generateFiles, writeFiles} from './generator';
import {getClient} from '@bgd-labs/rpc-env';
import {getBlockNumber} from 'viem/actions';

const program = new Command();

program
  .name('proposal-generator')
  .description('CLI to generate aave proposals')
  .version('1.0.0')
  .addOption(new Option('-f, --force', 'force creation (might overwrite existing files)'))
  .addOption(new Option('-p, --pools <pools...>').choices(POOLS))
  .addOption(new Option('-t, --title <string>', 'aip title'))
  .addOption(new Option('-a, --author <string>', 'author'))
  .addOption(new Option('-d, --discussion <string>', 'forum link'))
  .addOption(new Option('-c, --configFile <string>', 'path to config file'))
  .allowExcessArguments(false)
  .parse(process.argv);

let options = program.opts<Options>();
let poolConfigs: PoolConfigs = {};

const FEATURE_MODULES_V3 = [
  rateUpdatesV3,
  capsUpdates,
  collateralsUpdates,
  lstPriceCapsUpdates,
  stablePriceCapsUpdates,
  eModeUpdates
];

async function generateDeterministicPoolCache(pool: PoolIdentifier): Promise<PoolCache> {
  const chain = getPoolChain(pool);
  const client = getClient(CHAIN_TO_CHAIN_ID[chain], {});
  return {blockNumber: Number(await getBlockNumber(client))};
}

async function fetchPoolOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const features = await checkbox({
    message: `What do you want to do on ${pool}?`,
    choices: FEATURE_MODULES_V3.map((m) => ({value: m.value, name: m.description})),
  });
  for (const feature of features) {
    const module = FEATURE_MODULES_V3.find((m) => m.value === feature)!;
    poolConfigs[pool]!.configs[feature] = await module.cli({
      options,
      pool,
      cache: poolConfigs[pool]!.cache,
    });
    poolConfigs[pool]!.artifacts.push(
      module.build({
        options,
        pool,
        cfg: poolConfigs[pool]!.configs[feature],
        cache: poolConfigs[pool]!.cache,
      }),
    );
  }
}

if (options.configFile) {
  const {config: cfgFile}: {config: ConfigFile} = await import(
    path.join(process.cwd(), options.configFile)
  );
  options = {...options, ...cfgFile.rootOptions};
  poolConfigs = cfgFile.poolOptions as any;
  for (const pool of options.pools) {
    if (poolConfigs[pool]) {
      poolConfigs[pool]!.artifacts = [];
      for (const feature of Object.keys(poolConfigs[pool]!.configs)) {
        const module = FEATURE_MODULES_V3.find((m) => m.value === feature)!;
        poolConfigs[pool]!.artifacts.push(
          module.build({
            options,
            pool,
            cfg: poolConfigs[pool]!.configs[feature],
            cache: poolConfigs[pool]!.cache,
          }),
        );
      }
    } else {
      await fetchPoolOptions(pool);
    }
  }
} else {
  options.pools = await checkbox({
    message: 'Chains this proposal targets',
    choices: POOLS.map((v) => ({name: v, value: v})),
    required: true,
  });

  if (!options.title) {
    options.title = await input({
      message:
        'Short title of your steward update that will be used as contract name (please refrain from including author or date)',
      validate(input) {
        if (input.length == 0) return "Your title can't be empty";
        // this is no exact math
        // fully qualified identifiers are not allowed to be longer then 300 chars on etherscan api
        // the path is roughly src(3)/date(8)_title/title_date(8):title_date(8), so 3 + 3*8 + 3 title.length
        // so 80 sounds like a reasonable upper bound to stay below 300 character limit
        if (input.trim().length > 80) return 'Your title is to long';
        return true;
      },
    });
  }
  options.shortName = pascalCase(options.title);
  options.date = getDate();

  if (!options.author) {
    options.author = await input({
      message: 'Author of your proposal',
      validate(input) {
        if (input.length == 0) return "Your author can't be empty";
        return true;
      },
    });
  }

  if (!options.discussion) {
    options.discussion = await input({
      message: 'Link to forum discussion',
    });
  }

  for (const pool of options.pools) {
    await fetchPoolOptions(pool);
  }
}

try {
  const files = await generateFiles(options, poolConfigs);
  await writeFiles(options, files);
} catch (e) {
  console.log(JSON.stringify({options, poolConfigs}, null, 2));
  throw e;
}
