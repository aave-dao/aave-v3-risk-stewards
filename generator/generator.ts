import fs from 'fs';
import path from 'path';
import {generateContractName, generateFolderName} from './common';
import {proposalTemplate} from './templates/proposal.template';
import {confirm} from '@inquirer/prompts';
import {ConfigFile, Options, PoolConfigs, PoolIdentifier, Files} from './types';
import prettier from 'prettier';

const prettierSolCfg = await prettier.resolveConfig('foo.sol');
const prettierTsCfg = await prettier.resolveConfig('foo.ts');

/**
 * Generates all the file contents for aip/tests/payloads & script
 * @param options
 * @param poolConfigs
 * @returns
 */
export async function generateFiles(options: Options, poolConfigs: PoolConfigs): Promise<Files> {
  const jsonConfig = await prettier.format(
    `import {ConfigFile} from '../../generator/types';
    export const config: ConfigFile = ${JSON.stringify({
      rootOptions: options,
      poolOptions: (Object.keys(poolConfigs) as PoolIdentifier[]).reduce((acc, pool) => {
        acc[pool] = {configs: poolConfigs[pool]!.configs, cache: poolConfigs[pool]!.cache};
        return acc;
      }, {}),
    } as ConfigFile)}`,
    {...prettierTsCfg, filepath: 'foo.ts'},
  );

  async function createPayload(options: Options, pool: PoolIdentifier) {
    const contractName = generateContractName(options, pool);

    return {
      pool,
      payload: await prettier.format(proposalTemplate(options, poolConfigs[pool]!, pool), {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      contractName: contractName,
    };
  }

  return {
    jsonConfig,
    payloads: await Promise.all(options.pools.map((pool) => createPayload(options, pool))),
  };
}

async function askBeforeWrite(options: Options, path: string, content: string) {
  if (!options.force && fs.existsSync(path)) {
    const currentContent = fs.readFileSync(path, {encoding: 'utf8'});
    // skip if content did not change
    if (currentContent === content) return;
    const force = await confirm({
      message: `A file already exists at ${path} do you want to overwrite`,
      default: false,
    });
    if (!force) return;
  }
  fs.writeFileSync(path, content);
}

/**
 * Writes the files according to defined folder/file format
 * @param options
 * @param param1
 */
export async function writeFiles(options: Options, {jsonConfig, payloads}: Files) {
  const baseName = generateFolderName(options);
  const baseFolder = path.join(process.cwd(), 'src/contracts/updates/', baseName);

  if (fs.existsSync(baseFolder)) {
    if (!options.force && fs.existsSync(baseFolder)) {
      const force = await confirm({
        message: 'A proposal already exists at that location, do you want to continue?',
        default: false,
      });
      if (!force) return;
    }
  } else {
    fs.mkdirSync(baseFolder, {recursive: true});
  }

  // write config
  await askBeforeWrite(options, path.join(baseFolder, 'config.ts'), jsonConfig);

  for (const {payload, contractName} of payloads) {
    await askBeforeWrite(
      options,
      path.join(baseFolder, `${contractName}.sol`),
      payload,
    );
  }
}
