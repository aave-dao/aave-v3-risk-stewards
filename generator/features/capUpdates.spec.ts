// sum.test.js
import {expect, describe, it} from 'vitest';
import {MOCK_OPTIONS, capUpdate} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';
import {capsUpdates} from './capsUpdates';

describe('feature: capUpdates', () => {
  it('should return reasonable code', () => {
    const output = capsUpdates.build({
      options: MOCK_OPTIONS,
      pool: 'AaveV3Ethereum',
      cfg: capUpdate,
      cache: {blockNumber: 42},
    });
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      ['AaveV3Ethereum']: {
        artifacts: [
          capsUpdates.build({
            options: {...MOCK_OPTIONS, pools: ['AaveV3Ethereum']},
            pool: 'AaveV3Ethereum',
            cfg: capUpdate,
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.CAPS_UPDATE]: capUpdate},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles({...MOCK_OPTIONS, pools: ['AaveV3Ethereum']}, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
