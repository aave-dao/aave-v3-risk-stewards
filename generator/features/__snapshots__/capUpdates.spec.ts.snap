// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: capUpdates > should properly generate files 1`] = `
{
  "jsonConfig": "import {ConfigFile} from '../../generator/types';
export const config: ConfigFile = {
  rootOptions: {
    pools: ['AaveV3Ethereum'],
    title: 'test',
    shortName: 'Test',
    date: '20231023',
    author: 'test',
    discussion: 'test',
  },
  poolOptions: {
    AaveV3Ethereum: {
      configs: {CAPS_UPDATE: [{asset: 'WETH', supplyCap: '10000', borrowCap: '5000'}]},
      cache: {blockNumber: 42},
    },
  },
};
",
  "payloads": [
    {
      "contractName": "AaveV3Ethereum_Test_20231023",
      "payload": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
import {RiskStewardsEthereum} from '../../../../scripts/networks/RiskStewardsEthereum.s.sol';
import {IAaveV3ConfigEngine} from 'aave-v3-periphery/contracts/v3-config-engine/IAaveV3ConfigEngine.sol';

/**
 * @title test
 * @author test
 * - discussion: test
 * - deploy-command: make run-script contract=src/contracts/updates/20231023_AaveV3Ethereum_Test/AaveV3Ethereum_Test_20231023.sol:AaveV3Ethereum_Test_20231023 network=mainnet broadcast=false
 */
contract AaveV3Ethereum_Test_20231023 is RiskStewardsEthereum {
  function name() public pure override returns (string memory) {
    return 'AaveV3Ethereum_Test_20231023';
  }

  function capsUpdates() public pure override returns (IAaveV3ConfigEngine.CapsUpdate[] memory) {
    IAaveV3ConfigEngine.CapsUpdate[] memory capsUpdate = new IAaveV3ConfigEngine.CapsUpdate[](1);

    capsUpdate[0] = IAaveV3ConfigEngine.CapsUpdate({
      asset: AaveV3EthereumAssets.WETH_UNDERLYING,
      supplyCap: 10_000,
      borrowCap: 5_000
    });

    return capsUpdate;
  }
}
",
      "pool": "AaveV3Ethereum",
    },
  ],
}
`;

exports[`feature: capUpdates > should return reasonable code 1`] = `
{
  "code": {
    "fn": [
      "function capsUpdates() public pure override returns (IAaveV3ConfigEngine.CapsUpdate[] memory) {
          IAaveV3ConfigEngine.CapsUpdate[] memory capsUpdate = new IAaveV3ConfigEngine.CapsUpdate[](1);

          capsUpdate[0] = IAaveV3ConfigEngine.CapsUpdate({
               asset: AaveV3EthereumAssets.WETH_UNDERLYING,
               supplyCap: 10_000,
               borrowCap: 5_000
             });

          return capsUpdate;
        }",
    ],
  },
}
`;
